packages/opentelemetry-plugin-user-interaction/src/userInteraction.ts

```diff
@@ -0,0 +1,457 @@
+   * @param task
+   * @param currentZone
+   * @private
+   */
+  private _shouldCountTask(task: AsyncTask, currentZone: Zone): boolean {
+    if (task._zone) {
+      currentZone = task._zone;
+    }
+    if (!currentZone || !task.data || task.data.isPeriodic) {
+      return false;
+    }
+    const currentSpan = currentZone.get(ZONE_SCOPE_KEY);
+    if (!currentSpan) {
+      return false;
+    }
+    if (!this._uiMem.get(currentSpan)) {
+      return false;
+    }
+    return task.type === 'macroTask' || task.type === 'microTask';
+  }
+
+  /**
+   * Will try to end span when such span still exists.
+   * @param span
+   * @param endTime
+   * @private
+   */
+  private _tryToEndSpan(span: types.Span, endTime?: types.HrTime) {
+    if (span) {
+      const uiMem = this._uiMem.get(span);
+      if (uiMem) {
+        span.end(endTime);
+        this._uiMem.delete(span);
+      }
+    }
+  }
+
+  /**
+   * implements patch function
+   */
+  protected patch() {
+    const _window = window as any;
+    this._logger.debug(
+      'applying patch to',
+      this.moduleName,
+      this.version,
+      'zone:',
+      !!_window.Zone
+    );
+    if (this.isZoneAvailable()) {
+      const ZoneWithPrototype = (_window.Zone as unknown) as ZoneTypeWithPrototype;
+      if (isWrapped(ZoneWithPrototype.prototype.runTask)) {
+        shimmer.unwrap(ZoneWithPrototype.prototype, 'runTask');
+        this._logger.debug('removing previous patch from method runTask');
+      }
+      if (isWrapped(ZoneWithPrototype.prototype.scheduleTask)) {
+        shimmer.unwrap(ZoneWithPrototype.prototype, 'scheduleTask');
+        this._logger.debug('removing previous patch from method scheduleTask');
+      }
+      if (isWrapped(ZoneWithPrototype.prototype.cancelTask)) {
+        shimmer.unwrap(ZoneWithPrototype.prototype, 'cancelTask');
+        this._logger.debug('removing previous patch from method cancelTask');
+      }
+
+      this._zonePatched = true;
+      shimmer.wrap(
+        ZoneWithPrototype.prototype,
+        'runTask',
+        this._patchZoneRunTask()
+      );
+      shimmer.wrap(
+        ZoneWithPrototype.prototype,
+        'scheduleTask',
+        this._patchZoneScheduleTask()
+      );
+      shimmer.wrap(
+        ZoneWithPrototype.prototype,
+        'cancelTask',
+        this._patchZoneCancelTask()
+      );
+    } else {
+      this._zonePatched = false;
+      if (isWrapped(HTMLElement.prototype.addEventListener)) {
+        shimmer.unwrap(HTMLElement.prototype, 'addEventListener');
+        this._logger.debug(
+          'removing previous patch from method addEventListener'
+        );
+      }
+      shimmer.wrap(
+        HTMLElement.prototype,
+        'addEventListener',
+        this._patchElement()
+      );
+    }
+
+    this._patchHistoryApi();
+    return this._moduleExports;
+  }
+
+  /**
+   * implements unpatch function
+   */
+  protected unpatch() {
+    const _window = window as any;
```

Comment:
Similar comment here, can you make this an interface and do something like `window as WindowWithZone`?